### ISSUE TITLE
Block patterns with unregistered blocks should not be available within the Pattern Library

### ISSUE THREAD

**Issue by richtabor** [WordPress/gutenberg#31706]
## Description
Currently block patterns show up within the Pattern Library UI, regardless of if the pattern is insertable. Sure, one could add a broken pattern to a page, but its obviously a subpar experience when the "Your site doesn't support.." message shows up. 

I suggest that patterns with unregistered blocks are simply omitted from the Pattern Library, so that the experience of adding broken patterns to a site just doesn't occur. 

This will become an even bigger deal when the [WP Patterns Directory](https://wordpress.org/patterns/) lands - if third party blocks are allowed to be used within partners (as they should).

## Step-by-step reproduction instructions

1. Install a block/block plugin, such as [CoBlocks](https://wordpress.org/plugins/coblocks)
2. Create a pattern to your site using a block from it (I used the coblocks/gallery-carousel block in my example screenshot below). 
3. Deactivate the plugin (therefore unregistering the carousel block within the pattern).
2. Open the Pattern Library
3. See the broken pattern

## Expected behaviour
Unsupported patterns (patterns that would be broken if added to a page due to unregistered blocks) should not be available to insert via the Pattern Library.

## Actual behaviour
I can see and add broken patterns to my site.

## Screenshots
<img width="633" alt="Screen Shot 2021-05-11 at 8 29 15 AM" src="https://user-images.githubusercontent.com/1813435/117814988-fac11980-b232-11eb-91d8-2969afc0c1c8.png">

## WordPress information
- WordPress version: 5.7.1
- Gutenberg version: Not installed
- Are all plugins except Gutenberg deactivated? Yes (except for installing CoBlocks to make a pattern)
- Are you using a default theme (e.g. Twenty Twenty-One)? Yes

**Comment by jamiemarsland** [comment]
I agree with this 100% - the user experience would certainly be sub-par if patterns break. 

One thought is whether we could give the user feedback on why block patterns aren't showing up. i.e if a customer is used to seeing patterns on site A, and then suddenly doesn't see patterns on site B they might get confused.  Maybe we could grey out patterns as a way of giving users feedback that the patterns are there, but they can't be used.

**Comment by richtabor** [comment]
Thanks @jamiemarsland. I'm not 100% sold on graying out/disabling patterns visually, but perhaps a category to view broken patterns?

**Comment by mtias** [comment]
The fundamental idea for this flow is that patterns using blocks from the block directory would allow side-loading those blocks at the time of insertion, the same way that when a block is missing the editor allows users to directly install those blocks. The discussions so far have tied this process with the block directory, so plugins that bundle blocks would need to have them published or linked to the block directory for it to work.

Most of the groundwork here is tied to the block directory itself (#23378) with patterns benefitting by proxy. Patterns that use non-core block also need to clearly list them in the directory. It's conceivable that in some setups it would be preferred to only see patterns that use core blocks or currently installed blocks.

cc @tellyworth @ryelle

**Comment by skorasaurus** [comment]
similarly which requested to not display blocks that are filtered out through allowed_block_types https://github.com/WordPress/gutenberg/issues/23275

**Comment by richtabor** [comment]
Related #30965

**Comment by richtabor** [comment]
> The fundamental idea for this flow is that patterns using blocks from the block directory would allow side-loading those blocks at the time of insertion

I think its an interesting concept. Alternatively, should the blocks be side-loaded if the corresponding pattern is registered? The initial previews would showcase the broken/unregistered block within the UI with this flow. 🤔

**Comment by chthonic-ds** [comment]
> The fundamental idea for this flow is that patterns using blocks from the block directory would allow side-loading those blocks at  the time of insertion, the same way that when a block is missing the editor allows users to directly install those blocks. 

Per https://github.com/WordPress/gutenberg/issues/31706#issuecomment-838808387, how will this work with themes utilising an allow-list via allowed_block_types? Is it intended this flow will override instructions to disallow specific blocks? If not, it seems reasonable to only show block patterns containing available blocks.

**Comment by Mamaduka** [comment]
The option to install missing blocks directly from an editor is excellent. But it's only available for admin users; everyone else will only see the "Keep as HTML" action.

A lot of blocks also come with custom styling, and without them, things might appear broken.

I like the idea of side-loading missing blocks. Is there an issue where this is discussed?

**Comment by draganescu** [comment]
I think that there should at least be a setting to toggle sideloading unregistered blocks on or off. Some editor implementations might choose on purpose to not support certain blocks and that should reflect in patterns that include them to not appear.

For instance, the widgets editor currently does not support block theme blocks and reusable blocks. It's debatable if adding entity saving to widget areas makes sense. Maybe it does, but if it doesn't we should still be able to use the patterns that contain the supported blocks.

**Comment by melcarthus** [comment]
I ran into the opposite issue. I found this thread because I am using a custom block which has a block template inside it.
I also use an allowed_block_types_all filter to only keep the custom and core blocks that I want.

However... As I mentioned I use a custom block that has a block template inside it. And this template has blocks that are not allowed. But I noticed that although you can't pick those core blocks from the inserter they were working perfectly fine inside the custom block. So I thought saving them as a pattern should be no problem. I was wrong. I was wondering why my custom pattern was not showing up, and after allowing all the core blocks that I was using inside the template voila.... the pattern tab appeared again and I could see my patterns.

1. So if there is only 1 block that is not allowed and all the other blocks in a template are, the pattern will not show...
2. You can use block templates inside a block and not have the blocks allowed and the block will still work but saving it as a pattern won't let you use it anymore.

A solution would be if allowed blocks was not so absolute. But more conditional. As far as I know the only condition is post type. But I would like it if blocks could also be allowed on a root/parent level.

I apologize if this is already possible. But the documentation does not make this clear, nor does it make clear that patterns disappear if the blocks inside it are not allowed).

**Comment by Mamaduka** [comment]
This was fixed via #65700. There's an issue with a request to revert to the old behavior. We can continue the discussion there.
